plugins/wallet_plugin/se_wallet.cpp

```diff
@@ -0,0 +1,372 @@
+      ECDSA_SIG_set0(sig, r, s);
+
+      public_key_data kd;
+      compact_signature compact_sig;
+      try {
+         kd = get_public_key_data(it->second);
+         compact_sig = signature_from_ecdsa(key, kd, sig, d);
+      } catch(chain::wallet_exception&) {
+         CFRelease(signature);
+         CFRelease(digestData);
+         throw;
+      }
+
+      CFRelease(signature);
+      CFRelease(digestData);
+
+      char serialized_signature[sizeof(compact_sig) + 1];
+      serialized_signature[0] = 0x01;
+      memcpy(serialized_signature+1, compact_sig.data, sizeof(compact_sig));
+
+      signature_type final_signature;
+      fc::datastream<const char *> ds(serialized_signature, sizeof(serialized_signature));
+      fc::raw::unpack(ds, final_signature);
+      return final_signature;
+   }
+
+   bool remove_key(string public_key) {
+      auto it = _keys.find(public_key_type{public_key});
+      if(it == _keys.end())
+         FC_THROW_EXCEPTION(chain::wallet_exception, "Given key to delete not found in Secure Enclave wallet");
+
+      promise<bool> prom;
+      future<bool> fut = prom.get_future();
+      macos_user_auth(auth_callback, &prom, CFSTR("remove a key from your EOSIO wallet"));
+      if(!fut.get())
+         FC_THROW_EXCEPTION(chain::wallet_invalid_password_exception, "Local user authentication failed");
+
+      CFDictionaryRef deleteDic = CFDictionaryCreate(nullptr, (const void**)&kSecValueRef, (const void**)&it->second, 1, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+      OSStatus ret = SecItemDelete(deleteDic);
+      CFRelease(deleteDic);
+
+      if(ret)
+         FC_THROW_EXCEPTION(chain::wallet_exception, "Failed to getremove key from Secure Enclave");
+
+      CFRelease(it->second);
+      _keys.erase(it);
+
+      return true;
+   }
+
+   ~se_wallet_impl() {
+      for(auto& k : _keys)
+         CFRelease(k.second);
+   }
+
+   map<public_key_type,SecKeyRef> _keys;
+   fc::ec_key key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+   bool locked = true;
+};
+
+static void check_signed() {
+   OSStatus is_valid{0};
+   pid_t pid = getpid();
+   SecCodeRef code = nullptr;
+   CFNumberRef pidnumber = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type, &pid);
+   CFDictionaryRef piddict = CFDictionaryCreate(kCFAllocatorDefault, (const void**)&kSecGuestAttributePid, (const void**)&pidnumber, 1, nullptr, nullptr);
+   if(!SecCodeCopyGuestWithAttributes(nullptr, piddict, kSecCSDefaultFlags, &code)) {
+      is_valid = SecCodeCheckValidity(code, kSecCSDefaultFlags, 0);
+      CFRelease(code);
+   }
+   CFRelease(piddict);
+   CFRelease(pidnumber);
+
+   if(is_valid != errSecSuccess) {
+      wlog("Application does not have a valid signature; Secure Enclave support disabled");
+      FC_THROW("");
+   }
+}
+
+}
+
+se_wallet::se_wallet() : my(new detail::se_wallet_impl()) {
+   detail::check_signed();
+
+   //How to figure out of SE is available?!
```

Comment:
I realize how incredibly bogus this is. Will buy lunch to someone who can point out a quality way of detecting Secure Enclave support (beyond attempting to create a SE key and looking for a particular return code... which also isn't defined AFAIK)

Reply:
According to the interwebs, only Macs with an IOKit registry value for "AppleEmbeddedOSSupportHost" have secure enclaves.  That name leads me to believe it will be valid even if there's such a thing as a new Macbook with no touch bar but still a secure enclave.

Reply:
Looks like this `AppleEmbeddedOSSupportHost` is not present on iMacPros, which I have verified have a Secure Enclave and work with this wallet code.